test/index.spec.js

```diff
@@ -0,0 +1,956 @@
+import {
+  describe,
+  beforeAll,
+  test,
+  beforeEach,
+  afterEach,
+  expect,
+  vi,
+} from 'vitest';
+import { parse, stringify } from 'better-serialize';
+import Enmap from '../src';
+import { it } from 'node:test';
+import { mkdir, rm } from 'fs/promises';
+import CustomError from '../src/error';
+
+describe('Enmap', () => {
+  process.setMaxListeners(100);
+
+  describe('can instantiate', () => {
+    test('should create an Enmap', () => {
+      const enmap = new Enmap({ inMemory: true });
+
+      expect(enmap).toBeInstanceOf(Enmap);
+    });
+
+    test('should create an Enmap w/ warning', () => {
+      const spy = vi.spyOn(console, 'warn');
+
+      const enmap = new Enmap({ name: '::memory::' });
+
+      expect(enmap).toBeInstanceOf(Enmap);
+      expect(spy).toHaveBeenCalledTimes(1);
+    });
+
+    test('should create an Enmap w/ custom serializing', () => {
+      const enmap = new Enmap({
+        inMemory: true,
+        serializer: JSON.stringify,
+        deserializer: JSON.parse,
+      });
+
+      expect(enmap).toBeInstanceOf(Enmap);
+    });
+
+    test('should close database on exit', () => {
+      let callback;
+      process.on = (event, cb) => {
+        if (event === 'exit') {
+          callback = cb;
+        }
+      };
+
+      const enmap = new Enmap({ inMemory: true });
+
+      callback();
+
+      expect(enmap.db.open).toBe(false);
+    });
+
+    test('should create a persistent Enmap w/ dir', async () => {
+      await mkdir('/tmp/data').catch(() => {});
+
+      const enmap = new Enmap({ name: 'test', dataDir: '/tmp/data' });
+
+      expect(enmap).toBeInstanceOf(Enmap);
+    });
+
+    test('should load a persistent Enmap w/ dir', () => {
+      const enmap = new Enmap({ name: 'test', dataDir: '/tmp/data' });
+
+      expect(enmap).toBeInstanceOf(Enmap);
+    });
+
+    test('should fail to create a persistent Enmap w/o dir', async () => {
+      expect(
+        () => new Enmap({ name: 'test', dataDir: '/tmp/data-not-found' }),
+      ).toThrow(TypeError);
+    });
+
+    test('should create/use data dir', async () => {
+      await rm('./data', { recursive: true }).catch(() => {});
+
+      const enmap = new Enmap({ name: 'test' });
+      const enmap2 = new Enmap({ name: 'test' });
+
+      expect(enmap).toBeInstanceOf(Enmap);
+      expect(enmap2).toBeInstanceOf(Enmap);
```

Comment:
why do we need 2 of those tests for the creation of the data dir? also, should we not actually check the datar dir exists here?

Reply:
The second construct will hit the existsSync case, so it's essentially doing that for us

Reply:
ah nice 👍🏻